csharp/src/Apache.Arrow/Arrays/PrimitiveArrayBuilder.cs

```diff
@@ -162,8 +188,8 @@ public TBuilder Swap(int i, int j)
         public TArray Build(MemoryAllocator allocator = default)
         {
             return Build(
-                ValueBuffer.Build(allocator), ArrowBuffer.Empty,
-                ValueBuffer.Length, 0, 0);
+                ValueBuffer.Build(allocator), ValidityBuffer.Build(allocator).ValueBuffer,
```

Comment:
I agree with the sentiment, but given that every call to `Append[Range]` also touches `ValidityBuffer` I think this could end up trading-off a modest memory utilization improvement for [a modest-to-significant compute overhead](https://gist.github.com/mrange/aa82d33e94ad76d0f33ed86e704d7492) depending on the desired level of thread-safety and a given application's access pattern.

That said, `Build` could easily pass an empty buffer if `NullCount` == 0. Then `ArrowStreamWriter` could check the buffer's span `IsEmpty` property, and if so, emit [NullType instead](https://issues.apache.org/jira/browse/ARROW-6643).

Reply:
> That said, Build could easily pass an empty buffer if NullCount == 0. 

Yes that would work well. It is what the C++ implementation does:

https://github.com/apache/arrow/blob/3064a27cf284726ce08b98b14725b2f9d7ef5ea2/cpp/src/arrow/array.cc#L59-L62

Reply:
As noted on another comment, I've pushed up this refactor.